README.md
# flutter_touch_spin_2

A re-upload of [flutter_touch_spin](https://pub.dev/packages/flutter_touch_spin).
  
Reason: to keep dependencies up to date with the latest version of Reactive Forms, to use in the reactive_touch_spin widget.

Simple number input spinner Widget for Flutter. Inspired by [Bootstrap Touchspin](https://www.virtuosoft.eu/code/bootstrap-touchspin/)

___
<img src="https://github.com/danvick/flutter_touch_spin/blob/master/example/assets/screenshot.png?raw=true" alt="Screenshot" width="200"/>

___

## Getting Started

```dart
TouchSpin(
    min: 5,
    max: 100,
    step: 5,
    value: 10,
    displayFormat: NumberFormat.currency(locale: 'en_US', symbol: '\$'),
    textStyle: const TextStyle(fontSize: 36),
    iconSize: 48.0,
    addIcon: const Icon(Icons.add_circle_outline),
    subtractIcon: const Icon(Icons.remove_circle_outline),
    iconActiveColor: Colors.green,
    iconDisabledColor: Colors.grey,
    iconPadding: const EdgeInsets.all(20),
    onChanged: (val){
        print(val);
    },
),
```

## Attributes

| Attribute           | Type                 | Default                                    |
| ------------------- | -------------------- | ------------------------------------------ |
| `value`             | `num`                | `1`                                        |
| `min`               | `num`                | `1`                                        |
| `max`               | `num`                | `9999999`                                  |
| `step`              | `num`                | `1`                                        |
| `displayFormat`     | `NumberFormat`       |                                            |
| `textStyle`         | `TextStyle`          | `TextStyle(fontSize: 24)`                  |
| `iconSize`          | `double`             | `24.0`                                     |
| `addIcon`           | `Icon`               | `Icon(Icons.add)`                          |
| `subtractIcon`      | `Icon`               | `Icon(Icons.remove)`                       |
| `iconActiveColor`   | `Color`              | `Theme.of(context).textTheme.button.color` |
| `iconDisabledColor` | `Color`              | `Theme.of(context).disabledColor`          |
| `iconPadding`       | `EdgeInsetsGeometry` | `EdgeInsets.all(4.0)`                      |
| `onChanged`         | `ValueChanged<num>`  |                                            |


CHANGELOG.md
## [0.0.1] - 15-Jul-2023
- Initial release.


test/flutter_touch_spin_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_touch_spin_2/flutter_touch_spin.dart';

void main() {
  testWidgets('Touch Spin', (WidgetTester tester) async {
    await tester.pumpWidget(
      const MaterialApp(
        home: Scaffold(body: TouchSpin()),
      ),
    );
  });
}


lib/src/touch_spin.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

class TouchSpin extends StatefulWidget {
  final num value;
  final num min;
  final num max;
  final num step;
  final double iconSize;
  final ValueChanged<num>? onChanged;
  final NumberFormat? displayFormat;
  final Icon subtractIcon;
  final Icon addIcon;
  final EdgeInsetsGeometry iconPadding;
  final TextStyle textStyle;
  final Color? iconActiveColor;
  final Color? iconDisabledColor;
  final bool enabled;

  const TouchSpin({
    Key? key,
    this.value = 1.0,
    this.onChanged,
    this.min = 1.0,
    this.max = 9999999.0,
    this.step = 1.0,
    this.iconSize = 24.0,
    this.displayFormat,
    this.subtractIcon = const Icon(Icons.remove),
    this.addIcon = const Icon(Icons.add),
    this.iconPadding = const EdgeInsets.all(4.0),
    this.textStyle = const TextStyle(fontSize: 24),
    this.iconActiveColor,
    this.iconDisabledColor,
    this.enabled = true,
  }) : super(key: key);

  @override
  _TouchSpinState createState() => _TouchSpinState();
}

class _TouchSpinState extends State<TouchSpin> {
  late num _value;

  bool get minusBtnDisabled =>
      _value <= widget.min ||
      _value - widget.step < widget.min ||
      !widget.enabled;

  bool get addBtnDisabled =>
      _value >= widget.max ||
      _value + widget.step > widget.max ||
      !widget.enabled;

  @override
  void initState() {
    super.initState();
    _value = widget.value;
  }

  @override
  void didUpdateWidget(TouchSpin oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (oldWidget.value != widget.value) {
      _value = widget.value;
      widget.onChanged?.call(widget.value);
    }
  }

  Color? _spinButtonColor(bool btnDisabled) => btnDisabled
      ? widget.iconDisabledColor ?? Theme.of(context).disabledColor
      : widget.iconActiveColor ?? Theme.of(context).textTheme.button?.color;

  void _adjustValue(num adjustment) {
    num newVal = _value + adjustment;
    setState(() => _value = newVal);
    widget.onChanged?.call(newVal);
  }

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      crossAxisAlignment: CrossAxisAlignment.center,
      children: <Widget>[
        IconButton(
          padding: widget.iconPadding,
          iconSize: widget.iconSize,
          color: _spinButtonColor(minusBtnDisabled),
          icon: widget.subtractIcon,
          onPressed: minusBtnDisabled ? null : () => _adjustValue(-widget.step),
        ),
        Text(
          widget.displayFormat?.format(_value) ?? _value.toString(),
          style: widget.textStyle,
          key: ValueKey(_value),
        ),
        IconButton(
          padding: widget.iconPadding,
          iconSize: widget.iconSize,
          color: _spinButtonColor(addBtnDisabled),
          icon: widget.addIcon,
          onPressed: addBtnDisabled ? null : () => _adjustValue(widget.step),
        ),
      ],
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_touch_spin_2/flutter_touch_spin.dart';
import 'package:intl/intl.dart';

void main() => runApp(const MyApp());

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  double value = 10;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter TouchSpin Example',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        brightness: Brightness.dark,
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(title: const Text('Flutter TouchSpin')),
        body: Center(
          child: Column(
            children: [
              TouchSpin(
                key: UniqueKey(),
                value: value,
                min: 5,
                max: 100,
                step: 5,
                displayFormat:
                    NumberFormat.currency(locale: 'en_US', symbol: '\$'),
                textStyle: const TextStyle(fontSize: 36),
                iconSize: 48.0,
                addIcon: const Icon(Icons.add_circle_outline),
                subtractIcon: const Icon(Icons.remove_circle_outline),
                iconActiveColor: Colors.green,
                iconDisabledColor: Colors.grey,
                iconPadding: const EdgeInsets.all(20),
                onChanged: (val) {
                  debugPrint(val.toString());
                },
                enabled: true,
              ),
              const Divider(),
              Row(
                children: [
                  Expanded(
                    child: TouchSpin(
                      value: 10,
                      min: 5,
                      max: 100,
                      step: 1,
                      textStyle: const TextStyle(fontSize: 18),
                      iconSize: 24.0,
                      addIcon: const Icon(Icons.arrow_circle_up),
                      subtractIcon: const Icon(Icons.arrow_circle_down),
                      iconActiveColor: Colors.blue,
                      iconDisabledColor: Colors.grey,
                      iconPadding: const EdgeInsets.all(20),
                      onChanged: (val) {
                        debugPrint(val.toString());
                      },
                      enabled: true,
                    ),
                  ),
                  const VerticalDivider(),
                  Expanded(
                    child: TouchSpin(
                      value: 10,
                      min: 5,
                      max: 100,
                      step: 5,
                      textStyle: const TextStyle(fontSize: 18),
                      iconSize: 24,
                      addIcon: const Icon(Icons.arrow_forward),
                      subtractIcon: const Icon(Icons.arrow_back),
                      iconActiveColor: Colors.red,
                      iconDisabledColor: Colors.grey,
                      iconPadding: const EdgeInsets.all(20),
                      onChanged: (val) {
                        debugPrint(val.toString());
                      },
                      enabled: true,
                    ),
                  ),
                ],
              ),
              ElevatedButton(
                onPressed: () {
                  setState(() => value = 25);
                },
                child: const Text('Set to \$25'),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
